File: ko2koc.rb
#!/usr/bin/env ruby

astrIDs = []
ahashMaps = []
hashMap = fGenes = strSpecies = nil
STDIN.each do |strLine|
	if( strLine =~ /^ENTRY\s+(\S+)/ )
		astrIDs.push( $1 )
		ahashMaps.push( hashMap = {} )
	elsif( strLine =~ /^GENES(.+)/ )
		strLine = $1
		fGenes = true; end
	if( !fGenes || ( strLine =~ /^\/\/\// ) )
		fGenes = false
		next; end

	strLine.strip!
	if( strLine =~ /^(\S{3}):\s+(.+)/ )
		strSpecies, strLine = $1, $2; end
	strLine.split( /\s+/ ).each do |strToken|
		astrGenes = ( strToken =~ /^(\S+)\((.+)\)$/ ) ? [$1] : [strToken] # ,$2
		astrGenes.each do |strGene|
			hashMap[strSpecies + "#" + strGene.upcase] = true; end; end; end

astrIDs.each_index do |i|
	if( !ahashMaps[i].empty? )
		puts( astrIDs[i] + "\t" + ahashMaps[i].keys.join( "\t" ) ); end; end

=begin
hashhashMap = {}
strID = fGenes = nil
STDIN.each do |strLine|
	if( strLine =~ /^ENTRY\s+(\S+)/ )
		strID = $1
	elsif( strLine =~ /^GENES(.+)/ )
		strLine = $1
		fGenes = true; end
	if( !fGenes || ( strLine =~ /^\/\/\// ) )
		fGenes = false
		next; end

	strLine.strip!
	if( strLine =~ /^\S{3}:\s+(.+)/ )
		strLine = $1; end
	strLine.split( /\s+/ ).each do |strToken|
		astrGenes = ( strToken =~ /^(\S+)\((.+)\)$/ ) ? [$1, $2] : [strToken]
		astrGenes.each do |strGene|
			strGene.upcase!
			if( !( hashMap = hashhashMap[strGene] ) )
				hashhashMap[strGene] = hashMap = {}; end
			hashMap[strID] = true; end; end; end

hashhashMap.each do |strFrom, hashTo|
	puts( strFrom + "\t" + hashTo.keys.join( "\t" ) ); end
=end

File: src/ko2koc.py
#!/usr/bin/env python

import re
import sys

astrIDs = []
asetMaps = []
setMap = fGenes = strSpecies = None
for strLine in sys.stdin:
	mtch = re.search( '^ENTRY\s+(\S+)', strLine )
	if mtch:
		astrIDs.append( mtch.group( 1 ) )
		setMap = set()
		asetMaps.append( setMap )
		continue
	mtch = re.search( '^GENES(.+)', strLine )
	if mtch:
		strLine = mtch.group( 1 )
		fGenes = True
	mtch = re.search( '^\S', strLine )
	if ( not fGenes ) or mtch:
		fGenes = False
		continue

	strLine = strLine.strip( )
	mtch = re.search( '^(\S{3}):\s+(.+)', strLine )
	if mtch:
		strSpecies, strLine = mtch.groups( )
	for strToken in re.split( '\s+', strLine ):
		mtch = re.search( '^(\S+)\((.+)\)$', strToken )
		astrGenes = [mtch.group( 1 ) if mtch else strToken]
		for strGene in astrGenes:
			setMap.add( strSpecies + "#" + strGene.upper( ) )

for i in range( len( astrIDs ) ):
	if asetMaps[i]:
		print( "\t".join( [astrIDs[i]] + list(asetMaps[i]) ) )
